### ISSUE TITLE
scalafmt can't parse new scala 3.6.2 new given synstax

### ISSUE THREAD

**Issue by lenguyenthanh** [scalameta/scalafmt#4621]
The changes is here: https://github.com/lichess-org/lila-fishnet/pull/376/commits/1163f62523bf027ebb0b9a28c6e0e4a78535ac63

And the error is below:

```
[error] stack trace is suppressed; run last app / Compile / scalafmt for the full output
[error] (app / Compile / scalafmt) org.scalafmt.sbt.ScalafmtSbtReporter$ScalafmtSbtError: scalafmt: /Users/tle/git/lichess/lila-fishnet/app/src/main/scala/http/MkHttpServer.scala:19: error: [dialect scala3] abstract givens cannot be anonymous
[error]   given Logger[IO] => MkHttpServer = new:
[error]         ^ [/xxx/lichess/lila-fishnet/app/src/main/scala/http/MkHttpServer.scala]
[error] Total time: 1 s, completed 11 Dec 2024, 7:02:14 pm
```

this pr in scalameta may fix this issue: https://github.com/scalameta/scalameta/pull/4068

**Comment by kitbellew** [comment]
> this pr in scalameta may fix this issue: [scalameta/scalameta#4068](https://github.com/scalameta/scalameta/pull/4068)

So why submit it, then, if you already know the answer?

**Comment by lenguyenthanh** [comment]
I think this would help other people if they have the same issue. Also for tracking/release new scalafmt accordingly. Feel free to close if you see this just a noise.

**Comment by Dessix** [comment]
Even with the referenced fix, 3.8.4-RC4 does not seem to accept the following SIP-64 syntax for named, non-parametric givens:
```scala
trait MyTrait
object MyTrait:
  given myGiven: cats.Show[MyTrait]:
    def show(x: MyTrait): String = "MyTrait"
```
Yet, ironically, accepts the following nonsense:
```scala
trait MyTrait
object MyTrait:
  given myGiven:[A] => cats.Show[MyTrait]:
    def show(x: MyTrait): String = "MyTrait"
```

**Comment by kitbellew** [comment]
> Even with the referenced fix, 3.8.4-RC4 does not seem to accept the following SIP-64 syntax for named, non-parametric givens:
> 
> ```scala
> trait MyTrait
> object MyTrait:
>   given myGiven: cats.Show[MyTrait]:
>     def show(x: MyTrait): String = "MyTrait"
> ```

Can you please submit a separate issue to the scalameta repository?

> Scalafmt also lacks the ability to comprehend [SIP-64's new Context-Bounded Type Members](https://docs3.scala-lang.org/sips/sips/typeclasses-syntax.html#summary).

What does that mean? Can you provide examples of what it cannot comprehend, in a separate issue?

**Comment by Dessix** [comment]
> What does that mean? Can you provide examples of what it cannot comprehend

```scala
trait Foo:
  type Key : cats.Show
  type Value : {cats.Show, cats.Traverse}
  def entries: Iterable[(Key, Value)]
```
Produces the following error with `scalafmt 3.8.4-RC4`:
```
org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /example/src/.../example.scala:4: error: [dialect scala3] `=', `>:', or `<:' expectedles formatted
  type Key : cats.Show
           ^: /example/src/.../example.scala
```

**Comment by kitbellew** [comment]
> > What does that mean? Can you provide examples of what it cannot comprehend
> 
> ```scala
> trait Foo:
>   type Key : cats.Show
>   type Value : {cats.Show, cats.Traverse}
>   def entries: Iterable[(Key, Value)]
> ```
> 
> Produces the following error with `scalafmt 3.8.4-RC4`:
> 
> ```
> org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /example/src/.../example.scala:4: error: [dialect scala3] `=', `>:', or `<:' expectedles formatted
>   type Key : cats.Show
>            ^: /example/src/.../example.scala
> ```

Once again, could you please submit a new issue into the scalameta repository? scalafmt is responsible for formatting, not for parsing.

**Comment by kitbellew** [comment]
> Even with the referenced fix, 3.8.4-RC4 does not seem to accept the following SIP-64 syntax for named, non-parametric givens:
> 
> ```scala
> trait MyTrait
> object MyTrait:
>   given myGiven: cats.Show[MyTrait]:
>     def show(x: MyTrait): String = "MyTrait"
> ```
> 
> ```scala
> trait MyTrait
> object MyTrait:
>   given myGiven: cats.Show[MyTrait] {
>     def show(x: MyTrait): String = "MyTrait"
>   }
> ```

I cannot reproduce the problem you are describing; scalameta has no problem parsing these two, and scalafmt formats them without issue.

Can you please submit a new issue, with all the necessary information (such as your config file etc.)?

**Comment by Dessix** [comment]
... My apologies; In splitting this out to a separate project, I just now realized that scalafmt can delegate to other versions of itself based on a `version` tag in `.scalafmt.conf`. Mine was targeting `3.7.15` despite using the `3.8.4-RC4` executable.

Only the [type-member-context-bound syntax](https://github.com/scalameta/scalafmt/issues/4621#issuecomment-2564292993) still fails:
```scala
@main def hello(): Unit =
  println("Hello world!")
  println(msg)

def msg = "I was compiled by Scala 3. :)"

trait MyTrait:
  type X : cats.Show
  type Y : {cats.Show, cats.Monoid}
```
With `.scalafmt.conf`:
```
version = "3.8.4-RC4"
runner.dialect = scala3
```
And `build.sbt`:
```scala
val scala3Version = "3.6.2"

lazy val root = project
  .in(file("."))
  .settings(
    name := "Scala 3 Project Template",
    version := "0.1.0-SNAPSHOT",
    scalaVersion := scala3Version,
    libraryDependencies ++= Seq(
      "org.scalameta" %% "munit" % "1.0.0" % Test,
      "org.typelevel" %% "cats-core" % "2.12.0",
      "org.typelevel" %% "cats-effect" % "3.5.5",
      "org.typelevel" %% "cats-effect-cps" % "0.4.0",
      "org.typelevel" %% "cats-laws" % "2.12.0"
    )
  )
```
I receive the following error:
```
scalafmt
/tmp/example-scalafmt-fail/scala-3-project-template/src/main/scala/Main.scala:8: error: [dialect scala3] `=', `>:', or `<:' expected
  type X : cats.Show
```

**Comment by kitbellew** [comment]
@Dessix excellent, could you create a separate issue, so we can track any fixes against it?

**Comment by Dessix** [comment]
@kitbellew done in scalameta/scalameta#4093.

**Comment by Dessix** [comment]
@kitbellew The above issue I filed appears to be closed, but `scalafmt` still produces the following error on this trait:
```scala
trait T[Y] {
  type X : T
}
```
With the error message:
```
org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /home/dessix/code/irreverence/scs/src/main/scala/scs/util/graph/Graphish.scala:9: error: [dialect scala3] `=', `>:', or `<:' expected31 todo=62
  type X : T
         ^: /my/file/path.scala
```

`type X <: Y : Z` also fails with the same error.

**Comment by kitbellew** [comment]
> [@kitbellew](https://github.com/kitbellew) The above issue I filed appears to be closed, but `scalafmt` @ 3.9.7 still produces the following error on this trait:
> 
> trait T[Y] {
>   type X : T
> }
> With the error message:
> 
> ```
> org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /home/dessix/code/irreverence/scs/src/main/scala/scs/util/graph/Graphish.scala:9: error: [dialect scala3] `=', `>:', or `<:' expected31 todo=62
>   type X : T
>          ^: /my/file/path.scala
> ```
> 
> `type X <: Y : Z` also fails with the same error.

could you please submit a separate issue? this has nothing to do with the given _syntax_.

**Comment by kitbellew** [comment]
> [@kitbellew](https://github.com/kitbellew) The above issue I filed appears to be closed, but `scalafmt` @ 3.9.7 still produces the following error on this trait:
...
> ```
> org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /home/dessix/code/irreverence/scs/src/main/scala/scs/util/graph/Graphish.scala:9: error: [dialect scala3] `=', `>:', or `<:' expected31 todo=62
> ```

this error is not from a recent version of scalameta, hence you must not be using a recent version of scalafmt. your `3.9.7` should be from `.scalafmt.conf`, please double-check.
